### PR TITLE
new feature: add sm3 test case

### PR DESCRIPTION
**Pull request by ymj-uos**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
What is this adding beyond the existing test cases in `test/recipes/30-test_evp_data/evpmd_sm3.txt` ?

**Comment by ymj-uos** [comment]
> What is this adding beyond the existing test cases in `test/recipes/30-test_evp_data/evpmd_sm3.txt` ?

In the openssl, both test/recipes/30-test_evp_data/evpciph_sm4.txt and test/sm4_internal_test.c are existed.
so, test/sm3_internal_test.c offer another testing method, which enrich the perfect testing.

**Comment by ymj-uos** [comment]
I have completed and signed the Agreement (ICLA), which could be submited by the email (yangmingjun@uniontech.com).
How long does it take to reply？

**Comment by InfoHunter** [comment]
Can an ICLA be accepted with a company email? (I noticed the author's email address is a company one from uniontech.com)

**Comment by InfoHunter** [comment]
> In the openssl, both test/recipes/30-test_evp_data/evpciph_sm4.txt and test/sm4_internal_test.c are existed.
so, test/sm3_internal_test.c offer another testing method, which enrich the perfect testing.

From my understandings, what @paulidale asked is something about the content of the test cases (say the input and the expected output) between those two text case files (the evp one and the internal_text one).

**Comment by mattcaswell** [comment]
> Can an ICLA be accepted with a company email? (I noticed the author's email address is a company one from uniontech.com)

Contributions that are on behalf of an employer must have *both* an ICLA or CCLA. Where an email address appears to be a corporate one then we will usually ask for either a CCLA to be sent or for confirmation that the contributions are entirely personal. In this case I see that we received the ICLA and asked this exact question but have yet to receive a response (hence the "cla required" label is still present).

**Comment by levitte** [comment]
As for @paulidale's question, I see value in directly testing the innermost implementation of an algorithm.
The EVP tests do more than that, they also test availability of the algorithm and all the public mechanisms to fetch them, control them etc

**Comment by ymj-uos** [comment]
table.customTableClassName {margin-bottom: 10px;border-collapse: collapse;display: table;}.customTableClassName td, .customTableClassName th {border: 1px solid #ddd;}Hi,I have send CCLA to your official email, which is in the attachment. thanks for your attention. From: "Richard Levitte ***@***.***>"To: "openssl/openssl ***@***.***>"CC: "ymj-uos ***@***.***>","Author ***@***.***>"Sent: 2021-06-08 11:55Subject: Re: [openssl/openssl] new feature: add sm3 test case (#14271)As for @paulidale's question, I see value in directly testing the innermost implementation of an algorithm. The EVP tests do more than that, they also test availability of the algorithm and all the public mechanisms to fetch them, control them etc—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by ymj-uos** [comment]
I have send CCLA to your official email, please check your email, the title is "Re:Re: Submit ICLA about github/openssl (Submit Corporate CLA)"

**Comment by paulidale** [comment]
There is no sign of the CCLA yet.

**Comment by ymj-uos** [comment]
Is Official seal？or others?
image

**Comment by ymj-uos** [comment]
image
image
image

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot

**Comment by mattcaswell** [comment]
CLA issues are now resolved. Ping for review.

File: test/sm3_internal_test.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/*
+ * Internal tests for the SM3 module.
+ */
+
+#include <string.h>
+#include <openssl/opensslconf.h>
+#include "testutil.h"
+
+#ifndef OPENSSL_NO_SM3
+#include "crypto/sm3.h"
```
Comment (paulidale) [inline test/sm3_internal_test.c:20]: There should be a space after the # here.

File: test/recipes/03-test_internal_sm3.t [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Licensed under the OpenSSL license (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+use strict;
+use OpenSSL::Test;              # get 'plan'
+use OpenSSL::Test::Simple;
+use OpenSSL::Test::Utils;
```
Comment (paulidale) [inline test/recipes/03-test_internal_sm3.t:13]:
```Perl
use OpenSSL::Test::Simple;
```
should be enough.  The rest should be removed.

File: test/sm3_internal_test.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    sm3_init(&ctx1);
+    sm3_update(&ctx1, input1, sizeof(input1));
+    sm3_final(md1, &ctx1);
+    if (!TEST_mem_eq(md1, SM3_DIGEST_LENGTH, expected1, SM3_DIGEST_LENGTH))
+        return 0;
+
+    sm3_init(&ctx2);
+    sm3_update(&ctx2, input2, sizeof(input2));
+    sm3_final(md2, &ctx2);
```
Comment (paulidale) [inline test/sm3_internal_test.c:70]: The return codes should be checked (even if they can't fail):
```c
if (!TEST_true(sm3_init(...))
        || !TEST_true(sm3_update(...))
        || !TEST_true(sm3_final(...))
        || !TEST_mem_eq(...))
    return 0;
```

Similarly above.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: test/recipes/03-test_internal_sm3.t [hunk 4]
```diff
@@ -0,0 +1,15 @@
+#! /usr/bin/env perl
+# Copyright 2017-2021 The OpenSSL Project Authors. All Rights Reserved.
+# Copyright 2017 [Ribose Inc.](https://www.ribose.com). All Rights Reserved.
```
Comment (InfoHunter) [inline test/recipes/03-test_internal_sm3.t:3]: Are you from  Ribose? Then you need to update the year to 2021?

Comment (ymj-uos) [reply test/recipes/03-test_internal_sm3.t:3]: wait ...

Comment (ymj-uos) [reply test/recipes/03-test_internal_sm3.t]: The Copyright has been updated!

File: test/recipes/03-test_internal_sm3.t [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#! /usr/bin/env perl
+# Copyright 2017-2021 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (InfoHunter) [inline test/recipes/03-test_internal_sm3.t:2]: This should also be 2021 only since this file is added in 2021

File: test/sm3_internal_test.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+#include <string.h>
+#include <openssl/opensslconf.h>
+#include "testutil.h"
+
+#ifndef OPENSSL_NO_SM3
+# include "internal/sm3.h"
+
+static int test_sm3(void)
+{
+    static const unsigned char input1 [] = {
```
Comment (InfoHunter) [inline test/sm3_internal_test.c:24]: there should be no space between variable name and `[`.

File: test/sm3_internal_test.c [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        0x61, 0x62, 0x63
+    };
+
+    /*
+     * This test vector comes from Example 1 (A.1) of GM/T 0004-2012
+     */
+    static const unsigned char expected1 [SM3_DIGEST_LENGTH] = {
```
Comment (InfoHunter) [inline test/sm3_internal_test.c:31]: Same as previous comment

File: test/recipes/03-test_internal_sm3.t (hunk 5 above)
Comment (ymj-uos) [reply test/recipes/03-test_internal_sm3.t]: It has been modified!

File: test/sm3_internal_test.c (hunk 6 above)
Comment (ymj-uos) [reply test/sm3_internal_test.c]: It has been modified!

File: test/sm3_internal_test.c (hunk 7 above)
Comment (ymj-uos) [reply test/sm3_internal_test.c]: It has been modified!

**Review by paulidale — APPROVED** [review]

**Review by InfoHunter — APPROVED** [review]

**Comment by InfoHunter** [comment]
I clicked the `Approve and Run` button to run CIs on this PR...

**Comment by paulidale** [comment]
Merged after a sqashing.

Thanks for the contribution.

**Pull request closed without merging**